### PR TITLE
Mov-17,18,19,20 + pull genre with eventListner

### PR DESCRIPTION
**Pull request by Sergey113222**
MOV-17,18,19,20

### CODE REVIEW

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+@Service
+@AllArgsConstructor
+public class MovieServiceImpl implements MovieService {
+
+    private final MovieRepository movieRepository;
+    private final MovieConverter movieConverter;
+
+    @Override
+    public Long addMovie(MovieDto movieDto) {
+        return movieRepository.save(movieConverter.toModel(movieDto)).getId();
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: add check on existing movie before trying to persist it

Comment (RomanKashenok) [reply src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: Query(select m from Movie m where m.externalId = :externalId") should be added to MovieRepository

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+@Service
+@AllArgsConstructor
+public class MovieServiceImpl implements MovieService {
+
+    private final MovieRepository movieRepository;
+    private final MovieConverter movieConverter;
+
+    @Override
+    public Long addMovie(MovieDto movieDto) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: addToFavouriteMovies

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import info.movietrash.cinemabase.model.User;
+import info.movietrash.cinemabase.repository.MovieRepository;
+import info.movietrash.cinemabase.service.MovieService;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
+public class MovieServiceImpl implements MovieService {
+
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:21]: User u = new User() - stubbed user

**Review by RomanKashenok — CHANGES_REQUESTED** [review]

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java (hunk 1 above)
Comment (Sergey113222) [reply src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: done

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java (hunk 2 above)
Comment (Sergey113222) [reply src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: rename all methods done

File: src/main/java/info/movietrash/cinemabase/converter/impl/ProfileConverterImpl.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package info.movietrash.cinemabase.converter.impl;
+
+import info.movietrash.cinemabase.converter.ProfileConverter;
+import info.movietrash.cinemabase.dto.ProfileDto;
+import info.movietrash.cinemabase.model.Profile;
+
+public class ProfileConverterImpl implements ProfileConverter {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/ProfileConverterImpl.java:10]: make it spring component

File: src/main/java/info/movietrash/cinemabase/converter/impl/UserConverterImpl.java [hunk 5]
```diff
@@ -14,17 +14,23 @@ public User toModel(UserDto userDto) {
             throw new IllegalArgumentException("Some of required fields is null: " + userDto);
         }
         User user = new User();
+        ProfileConverterImpl profileConverter = new ProfileConverterImpl();
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/UserConverterImpl.java]: inject it as component

File: src/main/java/info/movietrash/cinemabase/converter/impl/UserConverterImpl.java [hunk 6]
```diff
@@ -14,17 +14,23 @@ public User toModel(UserDto userDto) {
...
         user.setPassword(userDto.getPassword());
+        user.setRole(userDto.getRole());
+        user.setProfile(profileConverter.toModel(userDto.getProfileDto()));
         return user;
     }
 
     @Override
     public UserDto toDto(User user) {
         UserDto userDto = new UserDto();
+        ProfileConverterImpl profileConverter = new ProfileConverterImpl();
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/UserConverterImpl.java]: same here

File: src/main/java/info/movietrash/cinemabase/dto/MovieDto.java [hunk 7]
```diff
@@ -26,5 +26,5 @@
     private Boolean adult;
     @JsonProperty("genre_ids")
     private List<Integer> genreIds;
-
+    private UserDto user;
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/dto/MovieDto.java]: че у тебя в фильме делают узеры?

File: src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
+package info.movietrash.cinemabase.service.impl;
+
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:2]: what is it?

File: src/main/java/info/movietrash/cinemabase/service/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -16,6 +16,7 @@
     private final UserRepository userRepository;
     private final UserConverter userConverter;
 
+    // @Transactional
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/UserServiceImpl.java]: why not transactinal?

**Review by RomanKashenok — CHANGES_REQUESTED** [review]

File: src/main/java/info/movietrash/cinemabase/controller/GenreController.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package info.movietrash.cinemabase.controller;
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/controller/GenreController.java]: Remove it

File: src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                .build()
+                .toUri();
+
+        RequestEntity request = new RequestEntity(HttpMethod.GET, uri);
+        ResponseEntity<String> response = restTemplate.exchange(request, String.class);
+        String genresJson = response.getBody();
+        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        JsonNode responseBody = objectMapper.readTree(genresJson);
+        JsonNode resultsMassive = responseBody.path(JSON_NODE_STR);
+        List<GenreDto> jsonGenreList = objectMapper.readValue(resultsMassive.toString(), new TypeReference<List<GenreDto>>() {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java]: add catch clause

File: src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+                genreList.add(genre);
+            }
+            return genreList;
+        }
+        return null;
+    }
+
+    @Override
+    public List<GenreDto> toDto(List<Genre> genreList) {
+        if (genreList != null) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java]: remove null check (or add CollectionUtils.isEmpty() )

File: src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+public class GenreConverterImpl implements GenreConverter {
+    @Override
+    public List<Genre> toModel(List<GenreDto> genreDtoList) {
+        if (genreDtoList != null) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java]: the same

File: src/main/java/info/movietrash/cinemabase/converter/impl/MovieConverterImpl.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+import info.movietrash.cinemabase.converter.MovieConverter;
+import info.movietrash.cinemabase.dto.MovieDto;
+import info.movietrash.cinemabase.model.Movie;
+import org.springframework.stereotype.Component;
+
+@Component
+public class MovieConverterImpl implements MovieConverter {
+    @Override
+    public Movie toModel(MovieDto movieDto) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/MovieConverterImpl.java]: (@Nonnull MovieDto movieDto)

File: src/main/java/info/movietrash/cinemabase/converter/impl/ProfileConverterImpl.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+import info.movietrash.cinemabase.converter.ProfileConverter;
+import info.movietrash.cinemabase.dto.ProfileDto;
+import info.movietrash.cinemabase.model.Profile;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ProfileConverterImpl implements ProfileConverter {
+    @Override
+    public Profile toModel(ProfileDto profileDto) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/ProfileConverterImpl.java]: the samw

File: src/main/java/info/movietrash/cinemabase/dto/GenreDto.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package info.movietrash.cinemabase.dto;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Data;
+
+@Data
+public class GenreDto {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/dto/GenreDto.java]: ExternalGenreDto ?

File: src/main/java/info/movietrash/cinemabase/service/GenreService.java [hunk 17]
```diff
@@ -0,0 +1,10 @@
+package info.movietrash.cinemabase.service;
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/GenreService.java]: remove it

File: src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        });
+        if (genreRepository.count() == 0) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:49]: move it as first part of method

File: src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+    public GenreServiceImpl(GenreRepository genreRepository, RestTemplate restTemplate, ObjectMapper objectMapper, GenreConverter genreConverter) {
+        this.genreRepository = genreRepository;
+        this.restTemplate = restTemplate;
+        this.objectMapper = objectMapper;
+        this.genreConverter = genreConverter;
+    }
+
+    @Override
+    @EventListener(ApplicationReadyEvent.class)
+    public void reloadData() throws JsonProcessingException {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java]: getGenresExternal

File: src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+
+    public GenreServiceImpl(GenreRepository genreRepository, RestTemplate restTemplate, ObjectMapper objectMapper, GenreConverter genreConverter) {
+        this.genreRepository = genreRepository;
+        this.restTemplate = restTemplate;
+        this.objectMapper = objectMapper;
+        this.genreConverter = genreConverter;
+    }
+
+    @Override
+    @EventListener(ApplicationReadyEvent.class)
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java]: move EventListener to application start class

File: src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    @Value("${themoviedb.ord.api-key}")
+    private String apiKey;
+    @Value("${themoviedb.ord.scheme}")
+    private String scheme;
+    @Value("${themoviedb.ord.host}")
+    private String host;
+
+    private final GenreRepository genreRepository;
+    private final RestTemplate restTemplate;
+    private final ObjectMapper objectMapper;
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:36]: move to configuration and make it @Bean or initialize as final field

File: src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.http.RequestEntity;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.net.URI;
+import java.util.List;
+
+@Service
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/GenreServiceImpl.java:24]: @RequiredArgsConstructor

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+    public Long addToFavouriteMovies(MovieDto movieDto) {
+        Long userId = movieDto.getUser().getId();
+        Optional<User> user = userRepository.findById(userId);
+
+        if (movieRepository.findByExternalId(movieDto.getExternalId()) != null) {
+            return movieRepository.findByExternalId(movieDto.getExternalId()).getId();
+        }
+        Movie movie = movieConverter.toModel(movieDto);
+        movie.getUsers().add(user.get());
+        return movieRepository.save(movie).getId();
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: rewrite to save user_movie

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+@AllArgsConstructor
+public class MovieServiceImpl implements MovieService {
+
+    private final MovieRepository movieRepository;
+    private final MovieConverter movieConverter;
+    private final UserRepository userRepository;
+
+    @Transactional
+    @Override
+    public Long addToFavouriteMovies(MovieDto movieDto) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java:29]: fixe like on video

**Review by RomanKashenok — CHANGES_REQUESTED** [review]

**Comment by RomanKashenok** [comment]
Fix conflicts! (merge branche 'development' into current)

File: src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java [hunk 25]
```diff
@@ -11,34 +11,30 @@
...
-                Genre genre = new Genre();
-                genre.setExternalId(genreDto.getExternalId());
-                genre.setName(genreDto.getName());
-                genreList.add(genre);
-            }
-            return genreList;
+    public List<Genre> toModel(List<ExternalGenreDto> externalGenreDtoList) {
+
+        List<Genre> genreList = new ArrayList<>();
+        for (int i = 0; i < externalGenreDtoList.size(); i++) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java]: use Stream API here.
externalGenreDtoList.stream().map(....).collect(....)

File: src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java [hunk 26]
```diff
@@ -11,34 +11,30 @@
...
-                Genre genre = genreList.get(i);
-                GenreDto genreDto = new GenreDto();
-                genreDto.setExternalId(genre.getExternalId());
-                genreDto.setName(genre.getName());
-                genreDtoList.add(genreDto);
-            }
-            return genreDtoList;
+    public List<ExternalGenreDto> toDto(List<Genre> genreList) {
+
+        List<ExternalGenreDto> externalGenreDtoList = new ArrayList<>();
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/converter/impl/GenreConverterImpl.java]: same as previous

File: src/main/java/info/movietrash/cinemabase/exception/ErrorControllerAdvice.java [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+import com.fasterxml.jackson.core.JsonProcessingException;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+@Slf4j
+public class ErrorControllerAdvice {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/exception/ErrorControllerAdvice.java]: Rename it to AooExceptionHandler or something like that

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java [hunk 28]
```diff
@@ -19,21 +21,35 @@
...
     @Transactional
     @Override
     public Long addToFavouriteMovies(MovieDto movieDto) {
-        Long userId = movieDto.getUser().getId();
-        Optional<User> user = userRepository.findById(userId);
-
-        if (movieRepository.findByExternalId(movieDto.getExternalId()) != null) {
-            return movieRepository.findByExternalId(movieDto.getExternalId()).getId();
+        Long userId = 1L;     // in my db have created user with id = 1
+        User user = null;
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: unnecessary variable

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java [hunk 29]
```diff
@@ -19,21 +21,35 @@
...
+        Movie movie = null;
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: the same

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java [hunk 30]
```diff
@@ -19,21 +21,35 @@
...
+        Optional<User> userOptional = userRepository.findById(userId);
+        if (userOptional.isPresent()) {
+            user = userOptional.get();
+        }
+        Optional<Movie> movieOptional = Optional.ofNullable(movieRepository.findByExternalId(movieDto.getExternalMovieId()));
+        if (movieOptional.isPresent()) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: replace if-else with
Movie movie = movieOptional.getOrDefault(...)

File: src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public List<MovieDto> searchMoviesByName(SearchDto dto) throws JsonProcessingException {
+        URI uri = UriComponentsBuilder.newInstance()
+                .scheme(scheme)
+                .host(host)
+                .path("/3/search/movie")
+                .queryParam("api_key", apiKey)
+                .queryParam("query", dto.getQuery())
+                .queryParam("language", dto.getLang())
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java]: searchDto has only nullable field, so query and lang could be null

File: src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java [hunk 32]
```diff
@@ -0,0 +1,105 @@
...
+        });
+        return jsonToMoviesList;
+    }
+
+    @Override
+    public MovieDto searchMovieLatest() throws JsonProcessingException {
+        URI uri = UriComponentsBuilder.newInstance()
+                .scheme(scheme)
+                .host(host)
+                .path("/3/movie/latest")
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java]: move path to .properties file

File: src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java [hunk 33]
```diff
@@ -0,0 +1,105 @@
...
+        });
+        return jsonToMoviesList;
+    }
+
+    @Override
+    public List<MovieDto> searchMoviesPopular() throws JsonProcessingException {
+        URI uri = UriComponentsBuilder.newInstance()
+                .scheme(scheme)
+                .host(host)
+                .path("/3/movie/popular")
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java]: move path to .properties file

File: src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java [hunk 34]
```diff
@@ -0,0 +1,105 @@
...
+        this.restTemplate = restTemplate;
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public List<MovieDto> searchMoviesByName(SearchDto dto) throws JsonProcessingException {
+        URI uri = UriComponentsBuilder.newInstance()
+                .scheme(scheme)
+                .host(host)
+                .path("/3/search/movie")
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/impl/SearchServiceImpl.java]: move path to .properties file

**Review by RomanKashenok — CHANGES_REQUESTED** [review]
change it a little please :)

File: src/main/java/info/movietrash/cinemabase/controller/UserMovieController.java [hunk 35]
```diff
@@ -23,8 +23,8 @@ public MovieDto findFavouriteMovieById(@PathVariable("id") Long id) {
     }
 
     @PutMapping
-    public void updateFavouriteMovie(@RequestBody MovieDto movieDto) {
-        movieService.updateFavouriteMovie(movieDto);
+    public void updateFavouriteMovie(@RequestBody MovieDto movieDto, @RequestParam Long userMovieId) {
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/controller/UserMovieController.java]: userMovieId - это зачем? Почему одно в теле, а второе в параметрах?
Может лучше отправлять userId в pathVariable, а в теле - только dto?

File: src/main/java/info/movietrash/cinemabase/service/MovieService.java [hunk 36]
```diff
@@ -9,7 +9,7 @@
 
     MovieDto fetchFavouriteMovieById(Long id);
 
-    void updateFavouriteMovie(MovieDto movieDto);
+    void updateFavouriteMovie(MovieDto movieDto, Long userMovieId);
```
Comment (RomanKashenok) [inline src/main/java/info/movietrash/cinemabase/service/MovieService.java:12]: подумай над Dto для класса UserMovie - не придется передавать два параметра :)

File: src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java (hunk 30 above)
Comment (RomanKashenok) [reply src/main/java/info/movietrash/cinemabase/service/impl/MovieServiceImpl.java]: а почему все же не getOrDefault ?

**Review by RomanKashenok — APPROVED** [review]
Я бы еще поработал над ним, но апрув поставлю.

**Pull request merged**